File: frontend/src/app/worknote/worknote.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';

export interface WorkNote {
  id: number;
  note: string;
  createdAt: string;

  maintenanceTask: {
    id: number;
    title: string;
    description: string;
    status: string;
    priority: string;

    aircraft: {
      id: number;
      tailNumber: string;
      aircraftType: string;
    } | null;
  };

  technician: {
    id: number;
    firstName: string;
    lastName: string;
    employeeNumber: string;
    specialty: string;
    active: boolean;
  };
}

@Injectable({
  providedIn: 'root'
})
export class WorkNoteService {

  private readonly apiUrl =
    'http://localhost:8081/api/work-notes';

  constructor(private http: HttpClient) {}

  getAllWorkNotes(): Observable<WorkNote[]> {
    return this.http.get<WorkNote[]>(
      this.apiUrl
    );
  }

  getWorkNotesByMaintenanceTaskId(
    maintenanceTaskId: number
  ): Observable<WorkNote[]> {

    return this.http.get<WorkNote[]>(
      `${this.apiUrl}/maintenance-task/${maintenanceTaskId}`
    );
  }

  createWorkNote(
    maintenanceTaskId: number,
    technicianId: number,
    note: string
  ): Observable<WorkNote> {

    const params = new HttpParams()
      .set(
        'maintenanceTaskId',
        maintenanceTaskId.toString()
      )
      .set(
        'technicianId',
        technicianId.toString()
      )
      .set(
        'note',
        note
      );

    return this.http.post<WorkNote>(
      this.apiUrl,
      null,
      { params }
    );
  }

  deleteWorkNote(
    id: number
  ): Observable<void> {
    return this.http.delete<void>(
      `${this.apiUrl}/${id}`
    );
  }

}

File: backend/src/main/java/com/opstrack/worknote/WorkNoteService.java
package com.opstrack.worknote;

import com.opstrack.maintenance.MaintenanceTask;
import com.opstrack.maintenance.MaintenanceTaskRepository;
import com.opstrack.technician.Technician;
import com.opstrack.technician.TechnicianRepository;
import org.springframework.stereotype.Service;

import java.util.List;

@Service
public class WorkNoteService {

    private final WorkNoteRepository workNoteRepository;
    private final MaintenanceTaskRepository maintenanceTaskRepository;
    private final TechnicianRepository technicianRepository;

    public WorkNoteService(
            WorkNoteRepository workNoteRepository,
            MaintenanceTaskRepository maintenanceTaskRepository,
            TechnicianRepository technicianRepository
    ) {
        this.workNoteRepository = workNoteRepository;
        this.maintenanceTaskRepository = maintenanceTaskRepository;
        this.technicianRepository = technicianRepository;
    }

    public WorkNote createWorkNote(
            Long maintenanceTaskId,
            Long technicianId,
            String note
    ) {
        MaintenanceTask maintenanceTask =
                maintenanceTaskRepository.findById(maintenanceTaskId)
                        .orElseThrow(() -> new IllegalArgumentException(
                                "Maintenance task not found with id: "
                                        + maintenanceTaskId
                        ));

        Technician technician =
                technicianRepository.findById(technicianId)
                        .orElseThrow(() -> new IllegalArgumentException(
                                "Technician not found with id: "
                                        + technicianId
                        ));

        WorkNote workNote = new WorkNote(
                note,
                maintenanceTask,
                technician
        );

        return workNoteRepository.save(workNote);
    }

    public List<WorkNote> getAllWorkNotes() {
        return workNoteRepository.findAll();
    }

    public List<WorkNote> getWorkNotesByMaintenanceTaskId(
            Long maintenanceTaskId
    ) {
        return workNoteRepository.findByMaintenanceTaskId(
                maintenanceTaskId
        );
    }

    public void deleteWorkNote(Long id) {

        WorkNote workNote =
                workNoteRepository.findById(id)
                        .orElseThrow(() -> new IllegalArgumentException(
                                "Work note not found with id: " + id
                        ));

        workNoteRepository.delete(workNote);
    }

}